src/hotspot/share/tsan/tsanOopMapTable.cpp

```diff
@@ -0,0 +1,159 @@
+}
+
+TsanOopMapTable::~TsanOopMapTable() {
+  clear();
+}
+
+bool TsanOopMapTable::add_oop_with_size(oop obj, size_t size) {
+  TsanOopMapTableKey new_entry(obj);
+  bool added;
+  if (obj->fast_no_hash_check()) {
+    added = _table.put_when_absent(new_entry, size);
+  } else {
+    size_t* v = _table.put_if_absent(new_entry, size, &added);
+    *v = size;
+  }
+
+  if (added) {
+    if (_table.maybe_grow(true /* use_large_table_sizes */)) {
+      log_info(tsan)("TsanOopMapTable resize to %d, %d entries",
+                     _table.table_size(), _table.number_of_entries());
+    }
+  }
+  return added;
+}
+
+#ifdef ASSERT
+bool TsanOopMapTable::is_empty() {
+  assert(TsanOopMap_lock->is_locked(), "sanity check");
+  return _table.number_of_entries() == 0;
+}
+
+size_t TsanOopMapTable::find(oop obj) {
+  if (is_empty()) {
+    return 0;
+  }
+
+  if (obj->fast_no_hash_check()) {
+    return 0;
+  }
+
+  TsanOopMapTableKey item(obj);
+  size_t* size = _table.get(item);
+  return size == nullptr ? 0 : *size;
+}
+#endif
+
+// - Notify Tsan about freed objects.
+// - Colllect objects moved bt GC and add a PendingMove for each moved
+//   objects in a GrowableArray.
+void TsanOopMapTable::collect_moved_objects_and_notify_freed(
+         GrowableArray<TsanOopMapImpl::PendingMove> *moves,
+         char **src_low, char **src_high,
+         char **dest_low, char **dest_high,
+         int *n_downward_moves) {
+  struct IsDead {
+    GrowableArray<TsanOopMapImpl::PendingMove> *_moves;
+    char **_src_low;
+    char **_src_high;
+    char **_dest_low;
+    char **_dest_high;
+    int  *_n_downward_moves;
+    IsDead(GrowableArray<TsanOopMapImpl::PendingMove> *moves,
+           char **src_low, char **src_high,
+           char **dest_low, char **dest_high,
+           int  *n_downward_moves) : _moves(moves), _src_low(src_low), _src_high(src_high),
+                                     _dest_low(dest_low), _dest_high(dest_high),
+                                     _n_downward_moves(n_downward_moves) {}
+    bool do_entry(TsanOopMapTableKey& entry, size_t size) {
+      oop wh_obj = entry.object_no_keepalive();
+      if (wh_obj == nullptr) {
+        log_trace(tsan)("__tsan_java_free for " PTR_FORMAT "\n", cast_from_oop<uintx>(entry.obj()));
+        __tsan_java_free(cast_from_oop<char*>(entry.obj()), size * HeapWordSize);
+        entry.release_weak_handle();
+        return true;
+      } else if (wh_obj != entry.obj()) {
+        TsanOopMapImpl::PendingMove move =
+          {cast_from_oop<char*>(entry.obj()), cast_from_oop<char*>(wh_obj), size * HeapWordSize};
+        _moves->append(move);
+        *_src_low = MIN2(*_src_low, move.source_begin());
+        *_src_high = MAX2(*_src_high, move.source_end());
+        *_dest_low = MIN2(*_dest_low, move.target_begin());
+        *_dest_high = MAX2(*_dest_high, move.target_end());
+        if (*_dest_low < *_src_low) {
+          ++(*_n_downward_moves);
+        }
+
+        entry.update_obj();
```

Comment:
It's safe because the entry still points to the same object, so it would not affect the key's hash value. This occurs during a STW GC pause, and access to the table from other concurrent worker threads is protected from the TsanOopMap_lock.